### ISSUE TITLE
Possible memory leak - some users report crashing and device rebooting with banner ads

### ISSUE THREAD

**Issue by gkovac42** [bdlukaa/native_admob_flutter#74]
This plugin seems to suffer from the same issue as [admob_flutter](https://pub.dev/packages/admob_flutter). A growing number my users are reporting consistent crashing and even device rebooting after loading banner ads. Sorry I can't provide more feedback, as I haven't been able to reporduce it on any of my devices, but I've noticed it happens mostly on Samsung devices, if that's any help...

**Comment by bdlukaa** [comment]
This may be an issue with hybrid composition. Are any of your users using Android 19?

**Comment by gkovac42** [comment]
It seems quite frequent on Samsung Galaxy devices devices with Android 10 and 7, but is not limited to that. It was also reported on a few devices with Android 9...

**Comment by bdlukaa** [comment]
Can you provide any logs? With Crashlytics or something

**Comment by gkovac42** [comment]
No, not at the moment. Never had much issues with crashing before so I didn't implement it. Only some obscure crash logs from play console android vitals, but these don't tell me much at all... any idea what to look for?

**Comment by bdlukaa** [comment]
> Possible memory leak

Make sure to dispose all ad controllers you have

**Comment by gkovac42** [comment]
I am disposing all of them already.  I must say that most of my users don't have any issues with ads whatsoever. But certain small number of devices does get these crashes and reboots consistently... it seems like a device or hardware specific problem with whatever you're doining with the native code, rather than flutter implementation problem :/ 

Sorry I can't be of more help at the moment... I'll try to get some logs from the afflicted users.

**Comment by bdlukaa** [comment]
> but not while using google_mobile_ads (had to switch from those cause of lag) so that there could be a clue...

May be an issue with virtual display

https://flutter.dev/docs/development/platform-integration/platform-views?tab=android-platform-views-kotlin-tab#performance

**Comment by kreativityapps** [comment]
Seeing some memory leaks in Crashlytics, maybe this log might help:

> Fatal Exception: java.lang.OutOfMemoryError: pthread_create (1040KB stack) failed: Try again
       at java.lang.Thread.nativeCreate(Thread.java)
       at java.lang.Thread.start(Thread.java:733)
       at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:975)
       at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1382)
       at com.google.android.gms.internal.ads.zzbbv.execute(zzbbv.java:2)
       at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:118)
       at com.google.android.gms.internal.ads.zzeei.zza(zzeei.java)
       at com.google.android.gms.ads.internal.util.zzj.b(zzj.java:19)
       at com.google.android.gms.internal.ads.zzbav.a(zzbav.java:28)
       at com.google.android.gms.internal.ads.zzbid.a(zzbid.java:53)
       at com.google.android.gms.internal.ads.zzbid.a(zzbid.java:27)
       at com.google.android.gms.internal.ads.zzbid.zzA(zzbid.java)
       at com.google.android.gms.ads.internal.ClientApi.b(ClientApi.java:6)
       at com.google.android.gms.internal.ads.zzzr.a(zzzr.java:13)
       at com.google.android.gms.internal.ads.zzzx.zze(zzzx.java:11)
       at com.google.android.gms.internal.ads.zzzx.a(zzzx.java:76)
       at com.google.android.gms.internal.ads.zzzw.a(zzzw.java:6)
       at com.google.android.gms.ads.AdLoader$Builder.<init>(AdLoader.java:17)
       at com.bruno.native_admob_flutter.native.NativeAdmobController.a(NativeAdmobController.java:145)
       at com.bruno.native_admob_flutter.native.NativeAdmobController.a(NativeAdmobController.java:200)
       at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.a(MethodChannel.java:17)
       at io.flutter.embedding.engine.dart.DartMessenger.a(DartMessenger.java:57)
       at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:4)
       at android.os.MessageQueue.nativePollOnce(MessageQueue.java)
       at android.os.MessageQueue.next(MessageQueue.java:327)
       at android.os.Looper.loop(Looper.java:196)
       at android.app.ActivityThread.main(ActivityThread.java:6760)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:911)

**Comment by bdlukaa** [comment]
Some people have told me to use [learkcanary](https://github.com/square/leakcanary) to find where the leak is happening on the device. https://square.github.io/leakcanary/

**Comment by ThrowJojo** [comment]
Thanks for creating this plugin. I recently migrated from `google_mobile_ads`. Everything is going pretty smoothly, I am getting some rare crashes from Android 9 devices(100%) though, and it might be related to this because it's a stack overflow error and looks like it has something to do with Platform Views. I'll attach some logs:

```
Fatal Exception: java.lang.StackOverflowError: stack size 8MB
       at java.lang.reflect.Method.invoke(Method.java)
       at io.flutter.view.AccessibilityViewEmbedder$ReflectionAccessors.getChildId(AccessibilityViewEmbedder.java:532)
       at io.flutter.view.AccessibilityViewEmbedder$ReflectionAccessors.access$500(AccessibilityViewEmbedder.java:429)
       at io.flutter.view.AccessibilityViewEmbedder.addChildrenToFlutterNode(AccessibilityViewEmbedder.java:171)
       at io.flutter.view.AccessibilityViewEmbedder.convertToFlutterNode(AccessibilityViewEmbedder.java:145)
       at io.flutter.view.AccessibilityViewEmbedder.getRootNode(AccessibilityViewEmbedder.java:94)
       at io.flutter.view.AccessibilityBridge.createAccessibilityNodeInfo(AccessibilityBridge.java:564)
       at android.view.View.populateVirtualStructure(View.java:8582)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.populateVirtualStructure(View.java:8585)
       at android.view.View.onProvideVirtualStructureCompat(View.java:8003)
       at android.view.View.onProvideVirtualStructure(View.java:7984)
       at android.view.View.dispatchProvideStructureForAssistOrAutofill(View.java:8650)
       at android.view.View.dispatchProvideStructure(View.java:8597)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3473)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3546)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3546)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3546)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3546)
       at android.app.assist.AssistStructure$WindowNode.<init>(AssistStructure.java:524)
       at android.app.assist.AssistStructure.<init>(AssistStructure.java:2052)
       at android.app.ActivityThread.handleRequestAssistContextExtras(ActivityThread.java:3311)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1863)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:6894)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:883)
```

It looks like a similar thing happens with the Google Maps plugin according to this issue: https://github.com/flutter/flutter/issues/38063 (I'm not using the Google Maps plugin, but they talk about hybrid composition platform views in the issue). It also looks like 67% of my crashes are happening in the background, so maybe it's likely to happen when the back button is pressed or the user returns to the app?

**Comment by bdlukaa** [comment]
Unfortunately Flutter's platform views are pretty unstable.

The log @ThrowJojo related is not related to the plugin itself, but to platform views.

The log @kreativityapps realted may be related to the plugin, but it's hard to tell where the problem is exactly because the log tells where the problem is in the COMPILED java code, not on the actual kotlin code. It may be due device-related problems (too many apps open in the background). It should not impact the majority of devices.

The best advice I can give to you on this issue is to:

- Make sure to `dispose` all the `NativeAdController`s you have, if any.
- Make sure to use the latest Flutter stable version

**Comment by bdlukaa** [comment]
The problem itself is that Flutter really isn't ready for admob yet. We can't call methods from another thread. This would likely solve the memory leak issue.

**Comment by bdlukaa** [comment]
Hello, @ThrowJojo . In what Flutter version are you in? This is probably fixed on the latest stable :)

**Comment by ThrowJojo** [comment]
Hey @bdlukaa sorry for not responding to your initial comments and thanks for taking the time to respond to me. 

This is pretty much the only (really minor) issue I have had so I'm happy to just ignore it considering Google itself is having issues with the Maps plugin. This plugin(native_admob_flutter) is great and I was finally able to add banner ads back to my app thanks to it.

In regards to the latest stable version, I shipped a staged rollout of the app with 2.2.2 and I'm still seeing the error in Crashlytics(it seems to be less frequent but? 🤔). I have yet to try out 2.2.3 so I will try updating to that for my next deployment.

**Comment by bdlukaa** [comment]
I'm glad you like the plugin :)

Indeed the crash can be caused by anything else, and the banner ad be just a "push" for what happened.

I'll keep this issue open, but feel free to close it

**Comment by jangruenwaldt** [comment]
Hey, I get the same issue still. Flutter 2.2.2

Android 9, Moto G6 Play

`Fatal Exception: java.lang.StackOverflowError: stack size 8MB
       at java.lang.reflect.Method.invoke(Method.java)
       at io.flutter.view.AccessibilityViewEmbedder$ReflectionAccessors.getChildId(AccessibilityViewEmbedder.java:532)
       at io.flutter.view.AccessibilityViewEmbedder$ReflectionAccessors.access$500(AccessibilityViewEmbedder.java:429)
       at io.flutter.view.AccessibilityViewEmbedder.addChildrenToFlutterNode(AccessibilityViewEmbedder.java:171)
       at io.flutter.view.AccessibilityViewEmbedder.convertToFlutterNode(AccessibilityViewEmbedder.java:145)
       at io.flutter.view.AccessibilityViewEmbedder.getRootNode(AccessibilityViewEmbedder.java:94)
       at io.flutter.view.AccessibilityBridge.createAccessibilityNodeInfo(AccessibilityBridge.java:564)
       at android.view.View.populateVirtualStructure(View.java:8580)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.populateVirtualStructure(View.java:8583)
       at android.view.View.onProvideVirtualStructureCompat(View.java:8001)
       at android.view.View.onProvideVirtualStructure(View.java:7982)
       at android.view.View.dispatchProvideStructureForAssistOrAutofill(View.java:8648)
       at android.view.View.dispatchProvideStructure(View.java:8595)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3475)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3548)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3548)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3548)
       at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3548)
       at android.app.assist.AssistStructure$WindowNode.<init>(AssistStructure.java:519)
       at android.app.assist.AssistStructure.<init>(AssistStructure.java:2047)
       at android.app.ActivityThread.handleRequestAssistContextExtras(ActivityThread.java:3394)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1933)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:6923)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:870)`

I can post more details if needed.

**Comment by bdlukaa** [comment]
`Fatal Exception: java.lang.StackOverflowError: stack size 8MB at` probably means the user has too many apps open and the phone is full of memory (it happens to me sometimes). The crash is not related to the plugin nor to admob nor to flutter.

**Comment by jangruenwaldt** [comment]
> `Fatal Exception: java.lang.StackOverflowError: stack size 8MB at` probably means the user has too many apps open and the phone is full of memory (it happens to me sometimes). The crash is not related to the plugin nor to admob nor to flutter.

I dont really think so... see here, seems to be caused when e.g. loops occur in code: https://stackoverflow.com/questions/52077206/fatal-exception-java-lang-stackoverflowerror-stack-size-8mb-on-camera-permissi

**Comment by ThrowJojo** [comment]
> I'm glad you like the plugin :)
> 
> Indeed the crash can be caused by anything else, and the banner ad be just a "push" for what happened.
> 
> I'll keep this issue open, but feel free to close it

If I had made this issue I would be happy to close it, unfortunately I just tagged along after someone else made it 😂 My apologies.

I haven't deployed an upgrade with v2.2.3 but I will report back when I do. 

I can say that I agree that this is just an issue/limitation with Platform Views, here's another issue with the same Stacktrace(apparently this person isn't using the maps plugin though):

https://github.com/flutter/flutter/issues/80233

**Comment by bdlukaa** [comment]
> apparently this person isn't using the maps plugin though

The device probably ran out of memory and force-closed the app. This happens frequently on low-end devices

**Comment by ThrowJojo** [comment]
> > apparently this person isn't using the maps plugin though
> 
> The device probably ran out of memory and force-closed the app. This happens frequently on low-end devices

Yeah sorry @bdlukaa that part of my response was more for @jangruenwaldt, I was just trying to say that I think it's an issue with Platform Views in general and not something that is fixable on your end. 👍

**Comment by naamapps** [comment]
Hey @bdlukaa, I'm testing your package on a Galaxy S21 Ultra. 
I face a critical issue - after some time with the app, the phone freezes and I have to restart my phone.
This didn't happen until I added ads to the my app.
I'll try to catch some logs.

**Comment by kittz** [comment]
Using example code from this plugin on fresh restarted Asus Zenfone Max Pro M1 (Android 9) gives these errors (when I switch to other app and switch back to this app). 
But this plugin worked really good when I tried on HTC 10 (Android 6), and Emulator (Android SDK 30).

Really like this plugin because the official one (google mobile ads) has performance issues.. I hope you can fix it.

I came across this post https://github.com/flutter/flutter/issues/38063#issuecomment-760026571 maybe this can help? Sorry I'm just a beginner.

Error:
E/AndroidRuntime(25212): FATAL EXCEPTION: main
E/AndroidRuntime(25212): Process: com.example.native_admob_flutter_example, PID: 25212
E/AndroidRuntime(25212): java.lang.StackOverflowError: stack size 8MB
E/AndroidRuntime(25212): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(25212): 	at io.flutter.view.AccessibilityViewEmbedder$ReflectionAccessors.getChildId(AccessibilityViewEmbedder.java:532)
E/AndroidRuntime(25212): 	at io.flutter.view.AccessibilityViewEmbedder$ReflectionAccessors.access$500(AccessibilityViewEmbedder.java:429)
E/AndroidRuntime(25212): 	at io.flutter.view.AccessibilityViewEmbedder.addChildrenToFlutterNode(AccessibilityViewEmbedder.java:171)
E/AndroidRuntime(25212): 	at io.flutter.view.AccessibilityViewEmbedder.convertToFlutterNode(AccessibilityViewEmbedder.java:145)
E/AndroidRuntime(25212): 	at io.flutter.view.AccessibilityViewEmbedder.getRootNode(AccessibilityViewEmbedder.java:94)
E/AndroidRuntime(25212): 	at io.flutter.view.AccessibilityBridge.createAccessibilityNodeInfo(AccessibilityBridge.java:564)
E/AndroidRuntime(25212): 	at android.view.View.populateVirtualStructure(View.java:8585)
113E/AndroidRuntime(25212): 	at android.view.View.populateVirtualStructure(View.java:8588)
I/flutter_exampl(25212): Background concurrent copying GC freed 722869(39MB) AllocSpace objects, 1(896KB) LOS objects, 41% free, 33MB/57MB, paused 115us total 103.028ms
I/chatty  (25212): uid=10111(com.example.native_admob_flutter_example) identical 906 lines
118E/AndroidRuntime(25212): 	at android.view.View.populateVirtualStructure(View.java:8588)
E/AndroidRuntime(25212): 	at android.view.View.onProvideVirtualStructureCompat(View.java:8006)
E/AndroidRuntime(25212): 	at android.view.View.onProvideVirtualStructure(View.java:7987)
E/AndroidRuntime(25212): 	at android.view.View.dispatchProvideStructureForAssistOrAutofill(View.java:8653)
E/AndroidRuntime(25212): 	at android.view.View.dispatchProvideStructure(View.java:8600)
E/AndroidRuntime(25212): 	at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3473)
4E/AndroidRuntime(25212): 	at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3546)
E/AndroidRuntime(25212): 	at android.app.assist.AssistStructure$WindowNode.<init>(AssistStructure.java:519)
E/AndroidRuntime(25212): 	at android.app.assist.AssistStructure.<init>(AssistStructure.java:2047)
E/AndroidRuntime(25212): 	at android.app.ActivityThread.handleRequestAssistContextExtras(ActivityThread.java:3178)
E/AndroidRuntime(25212): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1765)
E/AndroidRuntime(25212): 	at android.os.Handler.dispatchMessage(Handler.java:106)
E/AndroidRuntime(25212): 	at android.os.Looper.loop(Looper.java:193)
E/AndroidRuntime(25212): 	at android.app.ActivityThread.main(ActivityThread.java:6692)
E/AndroidRuntime(25212): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(25212): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/AndroidRuntime(25212): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
I/Process (25212): Sending signal. PID: 25212 SIG: 9
Lost connection to device.
Exited (sigterm)

**Comment by jangruenwaldt** [comment]
> Using example code from this plugin on fresh restarted Asus Zenfone Max Pro M1 (Android 9) gives these errors (when I switch to other app and switch back to this app).
> But this plugin worked really good when I tried on HTC 10 (Android 6), and Emulator (Android SDK 30).
> 
> Really like this plugin because the official one (google mobile ads) has performance issues.. I hope you can fix it.
> 
> I came across this post [flutter/flutter#38063 (comment)](https://github.com/flutter/flutter/issues/38063#issuecomment-760026571) maybe this can help? Sorry I'm just a beginner.
> 
> Error:
> E/AndroidRuntime(25212): FATAL EXCEPTION: main
> E/AndroidRuntime(25212): Process: com.example.native_admob_flutter_example, PID: 25212
> E/AndroidRuntime(25212): java.lang.StackOverflowError: stack size 8MB
> E/AndroidRuntime(25212): at java.lang.reflect.Method.invoke(Native Method)
> E/AndroidRuntime(25212): at io.flutter.view.AccessibilityViewEmbedder$ReflectionAccessors.getChildId(AccessibilityViewEmbedder.java:532)
> E/AndroidRuntime(25212): at io.flutter.view.AccessibilityViewEmbedder$ReflectionAccessors.access$500(AccessibilityViewEmbedder.java:429)
> E/AndroidRuntime(25212): at io.flutter.view.AccessibilityViewEmbedder.addChildrenToFlutterNode(AccessibilityViewEmbedder.java:171)
> E/AndroidRuntime(25212): at io.flutter.view.AccessibilityViewEmbedder.convertToFlutterNode(AccessibilityViewEmbedder.java:145)
> E/AndroidRuntime(25212): at io.flutter.view.AccessibilityViewEmbedder.getRootNode(AccessibilityViewEmbedder.java:94)
> E/AndroidRuntime(25212): at io.flutter.view.AccessibilityBridge.createAccessibilityNodeInfo(AccessibilityBridge.java:564)
> E/AndroidRuntime(25212): at android.view.View.populateVirtualStructure(View.java:8585)
> 113E/AndroidRuntime(25212): at android.view.View.populateVirtualStructure(View.java:8588)
> I/flutter_exampl(25212): Background concurrent copying GC freed 722869(39MB) AllocSpace objects, 1(896KB) LOS objects, 41% free, 33MB/57MB, paused 115us total 103.028ms
> I/chatty (25212): uid=10111(com.example.native_admob_flutter_example) identical 906 lines
> 118E/AndroidRuntime(25212): at android.view.View.populateVirtualStructure(View.java:8588)
> E/AndroidRuntime(25212): at android.view.View.onProvideVirtualStructureCompat(View.java:8006)
> E/AndroidRuntime(25212): at android.view.View.onProvideVirtualStructure(View.java:7987)
> E/AndroidRuntime(25212): at android.view.View.dispatchProvideStructureForAssistOrAutofill(View.java:8653)
> E/AndroidRuntime(25212): at android.view.View.dispatchProvideStructure(View.java:8600)
> E/AndroidRuntime(25212): at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3473)
> 4E/AndroidRuntime(25212): at android.view.ViewGroup.dispatchProvideStructure(ViewGroup.java:3546)
> E/AndroidRuntime(25212): at android.app.assist.AssistStructure$WindowNode.(AssistStructure.java:519)
> E/AndroidRuntime(25212): at android.app.assist.AssistStructure.(AssistStructure.java:2047)
> E/AndroidRuntime(25212): at android.app.ActivityThread.handleRequestAssistContextExtras(ActivityThread.java:3178)
> E/AndroidRuntime(25212): at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1765)
> E/AndroidRuntime(25212): at android.os.Handler.dispatchMessage(Handler.java:106)
> E/AndroidRuntime(25212): at android.os.Looper.loop(Looper.java:193)
> E/AndroidRuntime(25212): at android.app.ActivityThread.main(ActivityThread.java:6692)
> E/AndroidRuntime(25212): at java.lang.reflect.Method.invoke(Native Method)
> E/AndroidRuntime(25212): at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
> E/AndroidRuntime(25212): at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
> I/Process (25212): Sending signal. PID: 25212 SIG: 9
> Lost connection to device.
> Exited (sigterm)

Good point, and props to finding a potential fix. This should be tried for sure.

**Comment by jangruenwaldt** [comment]
@bdlukaa do you think its worth it to try the workaround posted by @kittz ?
